File: src/py/agent_studio/core/git_hygiene.py
"""Warn if the secret file (.env.local) is exposed to git.

A very common way keys leak is committing the env file. This surfaces that in
`doctor`. The pure ``env_file_warnings`` is unit-tested; ``check_env_file`` is a
thin wrapper that gathers git facts via subprocess (no-op outside a git repo).
"""

from __future__ import annotations

import subprocess
from pathlib import Path


def env_file_warnings(*, exists: bool, tracked: bool, ignored: bool, name: str = ".env.local") -> list[str]:
    if not exists:
        return []
    if tracked:
        return [
            f"{name} is TRACKED by git — your API key may get committed. "
            f"Fix: git rm --cached {name}  (and ensure it is in .gitignore)."
        ]
    if not ignored:
        return [f"{name} is not git-ignored — add it to .gitignore so secrets aren't committed."]
    return []


def _git(args: list[str], cwd: Path) -> int:
    try:
        return subprocess.run(
            ["git", *args], cwd=cwd, capture_output=True, timeout=5, check=False
        ).returncode
    except (OSError, subprocess.SubprocessError):
        return 1


def check_env_file(project_root: Path, name: str = ".env.local") -> list[str]:
    path = project_root / name
    if not path.exists():
        return []
    if _git(["rev-parse", "--is-inside-work-tree"], project_root) != 0:
        return []  # not a git repo — nothing to warn about
    tracked = _git(["ls-files", "--error-unmatch", name], project_root) == 0
    ignored = _git(["check-ignore", "-q", name], project_root) == 0
    return env_file_warnings(exists=True, tracked=tracked, ignored=ignored, name=name)

File: src/engine/core/gitHygiene.ts
// Warn if the secret file (.env.local) is exposed to git — a common way API
// keys leak. The pure `envFileWarnings` is unit-tested; `checkEnvFile` is a thin
// wrapper that gathers git facts via `git` (no-op outside a git repo).

import { spawnSync } from 'node:child_process';
import { existsSync } from 'node:fs';
import { join } from 'node:path';

export function envFileWarnings(o: {
  exists: boolean;
  tracked: boolean;
  ignored: boolean;
  name?: string;
}): string[] {
  const name = o.name ?? '.env.local';
  if (!o.exists) return [];
  if (o.tracked) {
    return [
      `${name} is TRACKED by git — your API key may get committed. ` +
        `Fix: git rm --cached ${name}  (and ensure it is in .gitignore).`,
    ];
  }
  if (!o.ignored) {
    return [`${name} is not git-ignored — add it to .gitignore so secrets aren't committed.`];
  }
  return [];
}

function git(args: string[], cwd: string): number {
  try {
    return spawnSync('git', args, { cwd, stdio: 'ignore', timeout: 5000 }).status ?? 1;
  } catch {
    return 1;
  }
}

export function checkEnvFile(projectRoot: string, name = '.env.local'): string[] {
  if (!existsSync(join(projectRoot, name))) return [];
  if (git(['rev-parse', '--is-inside-work-tree'], projectRoot) !== 0) return [];
  const tracked = git(['ls-files', '--error-unmatch', name], projectRoot) === 0;
  const ignored = git(['check-ignore', '-q', name], projectRoot) === 0;
  return envFileWarnings({ exists: true, tracked, ignored, name });
}
